primitives/src/transaction.rs

```diff
@@ -349,6 +351,388 @@ impl Encodable for Transaction {
     }
 }
 
+/// The decoder for the [`Transaction`] type.
+#[cfg(feature = "alloc")]
+pub struct TransactionDecoder {
+    state: TransactionDecoderState,
+}
+
+#[cfg(feature = "alloc")]
+impl TransactionDecoder {
+    /// Constructs a new [`TransactionDecoder`].
+    pub fn new() -> Self { Self { state: TransactionDecoderState::Version(VersionDecoder::new()) } }
+}
+
+#[cfg(feature = "alloc")]
+impl Default for TransactionDecoder {
+    fn default() -> Self { Self::new() }
+}
+
+/// The state of the transiting decoder.
+#[cfg(feature = "alloc")]
+pub enum TransactionDecoderState {
+    /// Decoding the transaction version.
+    Version(VersionDecoder),
+    /// Decoding the transaction inputs.
+    Inputs(Version, Attempt, VecDecoder<TxIn>),
+    /// Decoding the segwit flag.
+    SegwitFlag(Version),
+    /// Decoding the transaction outputs.
+    Outputs(Version, Vec<TxIn>, IsSegwit, VecDecoder<TxOut>),
+    /// Decoding the segwit transaction witnesses.
+    Witnesses(Version, Vec<TxIn>, Vec<TxOut>, Iteration, WitnessDecoder),
+    /// Decoding the transaction lock time.
+    LockTime(Version, Vec<TxIn>, Vec<TxOut>, LockTimeDecoder),
+    /// Done decoding the [`Transaction`].
+    Done(Transaction),
+    /// Temporary state during transitions, should never be observed.
+    Transitioning,
+}
+
+/// Boolean used to track number of times we have attempted to decode the inputs vector.
+#[cfg(feature = "alloc")]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Attempt {
+    /// First time reading inputs.
+    First,
+    /// Second time reading inputs.
+    Second,
+}
+
+/// Boolean used to track whether or not this transaction uses segwit encoding.
+#[cfg(feature = "alloc")]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum IsSegwit {
```

Comment:
In 15b4f4c43bc790e2e86797f5dee56f7af588ba54, another minor clarity idea for the state machine is to fold `Attempt` into `IsSegwit` by adding some sort of initial `Unknown` variant. Might make the `Inputs`/`SegwitFlag` dance a little clearer.

Reply:
So with all these types private, your idea (which I'm not exactly understanding right now) becomes an internal change, right? So I can punt it down the field?

Reply:
Yep, can punt.

And said another way, instead of having the `Attempt` enum for the `Inputs` state, just use the `IsSegwit` state.

```rust
enum IsSegwit {
    /// Yes so uses segwit encoding.
    Yes,
    /// No segwit flag, marker, or witnesses.
    No,
    /// Initial state, inputs have not been decoded yet.
    Unknown,
}
```

This isn't a massive change, just maps well in my mind to how there are really 3 Segwit states when decoding.

Reply:
Ah yeah that looks good.